### ISSUE TITLE
Initialize New Accounts with Nonce Equal to Block Number

### ISSUE THREAD

**Issue by shawntabrizi** [paritytech/substrate#8457]
There is a well known issue that Substrate accounts which are killed could have their txs replayed if the lifetime is infinite.

To help mitigate this, we want to initialize a user's nonce with the number of the current block.

This means that assuming that an account does on average less than one tx per block, that none of it's transactions could be replayed even if the account was killed.

This can also be used as a mechanism to determine how old an account is (relatively). For example, we could start a democracy vote where only users with nonce lower than the current block number could vote. This means that it would not be possible for someone to create a bunch of new accounts after the vote is announced to then influence the outcome.

It would of course be possible to create a bunch of sleeper accounts initialized with a low nonce, but such activity could be monitored and accounted for before the vote takes place. Anyway, this second half is just an idea...

**Comment by kevlu93** [comment]
I would be interested in helping out! New to the code base though so would need some pointers to get started.

**Comment by xlc** [comment]
This helps but immortal tx with nonce of 0 is still going to be replayable

**Comment by shawntabrizi** [comment]
> This helps but immortal tx with nonce of 0 is still going to be replayable

It would be simple enough to not allow nonce 0?

**Comment by Lohann** [comment]
Can I take this one? I just faced this issue trying to figure why two transactions in different blocks have the same hash and signature, block explores rely in this functionally for uniquely identify the transaction.

Tx hash: `0x313da69888b43486a5bf717958d02e7228467fa48654c8bd690fe38b56370113`

https://polkascan.io/polkadot/block/7911545
https://polkascan.io/polkadot/block/7854172

**Comment by shawntabrizi** [comment]
That problem is not the same thing as mentioned in this issue.

A tx hash is literally a hash of the tx input data, and can be the same if the same inputs are used AFAIK.

The unique identifier for a tx is (block number, extrinsic index).

**Comment by Lohann** [comment]
In my opinion, is better to disable the `Pallet::<T>::on_killed_account(who.clone())` functionality in the System Pallet until this issue if fixed, then create a migration which deletes all accounts without providers in the future. Because I consider the risk o replay attack worst than the risk of storage bloating.

https://github.com/paritytech/substrate/blob/polkadot-v0.9.13/frame/system/src/lib.rs#L1080-L1085

I don't think people are aware that if their account is deleted then re-created, they can suffer replay attacks.

**Comment by Lohann** [comment]
I noticed that fixing the nonce initialization is not trivial:
- https://github.com/paritytech/substrate/pull/8822
- https://github.com/paritytech/frontier/issues/445

So my suggestion is doing a band-aid solution to prevent replay attacks, maybe something like this:

[frame/system/src/lib.rs#L1080-L1085](https://github.com/paritytech/substrate/blob/polkadot-v0.9.13/frame/system/src/lib.rs#L1080-L1085)
```rust
(1, 0, 0) => {
	// No providers left (and no consumers) and no sufficients. Account dead.

	if account.nonce > 0 {
		Pallet::<T>::on_killed_account(who.clone());
		Ok(DecRefStatus::Reaped)
	} else {
		// Mark the account to be reaped in a future migration
	}
},
```

**Comment by shawntabrizi** [comment]
> same hash because the account reuse the same nonce and call the same extrinsic twice

This is not accurate. The tx hash does not include the nonce of the user or the account. Any two accounts will generate the same tx hash with the same tx and inputs. It is literally the hash of the tx bytes, not a unique identifier.

> I don't think people are aware that if their account is deleted then re-created, they can suffer replay attacks.

This is a completely documented and known thing with functions already built into polkadot to prevent it.

All extrinsics have a configurable transaction lifetime or mortality. You can learn more about that here: https://wiki.polkadot.network/docs/build-protocol-info#transaction-mortality

**Comment by leonardocustodio** [comment]
I've been working with polkadot for a few months now and I did not know the possibility of replaying transactions in that way.... I bet people that are not programmers or anything will never know as well...
If you don't think that's a serious security flaw, I'm not sure what is.

**Comment by shawntabrizi** [comment]
I am not sure what to respond here. It is a part of the protocol, I am sure you and I both don't know every detail of Polkadot, however this is a feature which exists and when used correctly, can prevent tx replays.

Every wallet creator worth their salt in the Polkadot ecosystem definitely makes sure to provide only mortal transactions by default.

**Comment by leonardocustodio** [comment]
Yes, I've read the documentation that you sent and it does say in there... But like you just said I don't know every detail, you don't know, a lot of people don't know. Why should substrate expect that everybody will know? And people that run their own nodes instead of using a wallet, will also know this detail?

I believe security flaws from the blockchain should be fixed on the blockchain.... And not asking the wallets to do things to avoid that or expecting everybody including people that don't use the wallet to know about the security flaw.

That's basically like saying: "there is a security issue in HSBC banking system.... but we will fix it in the mobile app not in the banking system itself"

I still agree with @Lohann that this is indeed a risk that should be fixed somehow.
